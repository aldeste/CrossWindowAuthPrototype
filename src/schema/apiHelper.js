/* @flow */
import DataLoader from "dataloader";
import chalk from "chalk";
import { Person as dbPerson, Planet as dbPlanet } from "../data";
import User, { type Viewer } from "../models/User";
import Planet from "../models/Planet";

import { type GraphQLObjectType } from "graphql";

export type Loaders = {
  Person: Function,
  Planet: Function
};

export type DataLoaders = {
  Person: DataLoader<*, *>,
  Planet: DataLoader<*, *>
};

/**
 * This resolves a query to each id, batching the array into a
 * single fetch. this function is memoized as a response to DataLoader,
 * meaning the same ID won't be fetched twice in the same query.
 */
export function getResolve(
  type: Class<*>,
  ids: Array<string>,
  options: Object = {}
): Promise<*> {
  return Promise.all(
    ids.map(id => {
      // We console log each request to ilustrate how dataloader works,
      // If the same reqouse is requisted twice durning the same request
      // it'll still only consol log once, since the result is memoized.
      console.log(
        chalk.bold.magenta(
          "Fetched",
          chalk.yellow.italic(type.name),
          "with id",
          chalk.yellow.italic(id),
          "from database"
        )
      );
      return type.scope("withIds").findById(id, options);
    })
  );
}

/**
 * A factory to create DataLoader memoized database calls. Best
 * practice is to create each per request, this will help if
 * same data is fetched multiple times durning the same request.
 */
export function createLoaders(Viewer: ?Viewer): Loaders {
  const loaders: DataLoaders = {
    Person: new DataLoader(ids => getResolve(dbPerson, ids)),
    Planet: new DataLoader(ids => getResolve(dbPlanet, ids))
  };

  return {
    Person: (id: string): Promise<?User> =>
      User.gen(Viewer || null, id, loaders),
    PersonConnection: ids => User.genMany(Viewer || null, ids, loaders),
    Planet: (id: string): Promise<?Object> =>
      Planet.gen(Viewer || null, id, loaders)
  };
}

/**
 * Given a type and ID, get the object with the ID.
 */
export async function getObjectFromTypeAndId(
  type: Object,
  id: string,
  context?: Object
): Promise<Object> {
  const Loaders = (context && context.loaders) || createLoaders();
  return await Loaders[type.name](id);
}

/**
 * Given a type, fetch all objects
 */
export async function getObjectsByType(
  type: GraphQLObjectType,
  args?: ?Object,
  context: Object,
  ids: Array<string>
): Promise<Object> {
  const Loaders = (context && context.loaders) || createLoaders();
  const objects = await Loaders[type.name + "Connection"](ids);

  return { objects, currentCount: objects.length };
}
