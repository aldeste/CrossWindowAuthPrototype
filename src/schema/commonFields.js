// @flow

/**
 * This file contains fields used by each GraphQL type. Instead
 * of defining them multiple times, they are defined here.
 */

import { GraphQLString, type GraphQLFieldConfig } from "graphql";

export function createdField(): GraphQLFieldConfig<*, *> {
  return {
    type: GraphQLString,
    description:
      "The ISO 8601 date format of the time that this resource was created.",
    resolve: current => current.created
  };
}

export function editedField(): GraphQLFieldConfig<*, *> {
  return {
    type: GraphQLString,
    description:
      "The ISO 8601 date format of the time that this resource was edited.",
    resolve: current => current.edited
  };
}
