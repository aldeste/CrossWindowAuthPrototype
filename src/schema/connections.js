/* @flow */
import {
  connectionFromArray,
  connectionArgs,
  connectionDefinitions
} from "graphql-relay";
import { getObjectsByType } from "./apiHelper";
import {
  type GraphQLObjectType,
  type GraphQLFieldConfig,
  GraphQLInt
} from "graphql";

export function connectionFromType(
  name: string,
  type: GraphQLObjectType,
  idField: string
): GraphQLFieldConfig<*, *> {
  const { connectionType } = connectionDefinitions({
    name,
    nodeType: type,
    resolveNode: edge => edge.node,
    connectionFields: () => ({
      currentCount: {
        type: GraphQLInt,
        resolve: conn => conn.currentCount,
        description: `A count of the total number of objects in this connection, ignoring pagination.
This allows a client to fetch the first five objects by passing "5" as the
argument to "first", then fetch the total count so it could display "5 of 83",
for example.`
      }
    })
  });
  return {
    type: connectionType,
    args: connectionArgs,
    resolve: async (currentType: Object, args: Object, ctx: Object) => {
      const { objects, currentCount } = await getObjectsByType(
        type,
        args,
        ctx,
        currentType[idField].map(x => x.id)
      );
      return {
        ...connectionFromArray(objects, args),
        currentCount
      };
    }
  };
}
