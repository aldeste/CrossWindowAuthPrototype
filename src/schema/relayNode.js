// @flow
import { fromGlobalId, nodeDefinitions } from "graphql-relay";

import PersonType from "./types/personType";
import PlanetType from "./types/planetType";

import { getObjectFromTypeAndId } from "./apiHelper";

export function idFetcher(
  globalId: string,
  context: Object
): Promise<Object> | null {
  const { type, id } = fromGlobalId(globalId);
  const resolve = (type: Object) => getObjectFromTypeAndId(type, id, context);
  if (type === "people") {
    return resolve(PersonType);
  }
  if (type === "planets") {
    return resolve(PlanetType);
  }
  return null;
}

export function typeResolver(object: { GraphQLType?: Object }): Object | null {
  if (object.GraphQLType === PersonType.name) {
    return PersonType;
  }
  if (object.GraphQLType === PlanetType.name) {
    return PlanetType;
  }
  return null;
}

const { nodeInterface, nodeField } = nodeDefinitions(idFetcher, typeResolver);

export { nodeInterface, nodeField };
