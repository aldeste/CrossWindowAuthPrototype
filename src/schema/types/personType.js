// @flow
import { GraphQLInt, GraphQLObjectType, GraphQLString } from "graphql";
import { globalIdField } from "graphql-relay";

import { nodeInterface } from "../relayNode";
import { createdField, editedField } from "../commonFields";

/**
 * The GraphQL type equivalent of the People resource
 */
const PersonType = new GraphQLObjectType({
  name: "Person",
  description: "An individual person or character within the Star Wars universe.",
  fields: () => ({
    name: {
      type: GraphQLString,
      description: "The name of this person."
    },
    birthYear: {
      type: GraphQLString,
      resolve: person => person.birthYear,
      description: `The birth year of the person, using the in-universe standard of BBY or ABY -
Before the Battle of Yavin or After the Battle of Yavin. The Battle of Yavin is
a battle that occurs at the end of Star Wars episode IV: A New Hope.`
    },
    eyeColor: {
      type: GraphQLString,
      resolve: person => person.eyeColor,
      description: `The eye color of this person. Will be "unknown" if not known or "n/a" if the
person does not have an eye.`
    },
    gender: {
      type: GraphQLString,
      description: `The gender of this person. Either "Male", "Female" or "unknown",
"n/a" if the person does not have a gender.`
    },
    hairColor: {
      type: GraphQLString,
      resolve: person => person.hairColor,
      description: `The hair color of this person. Will be "unknown" if not known or "n/a" if the
person does not have hair.`
    },
    height: {
      type: GraphQLInt,
      description: "The height of the person in centimeters."
    },
    mass: {
      type: GraphQLInt,
      description: "The mass of the person in kilograms."
    },
    skinColor: {
      type: GraphQLString,
      resolve: person => person.skinColor,
      description: "The skin color of this person."
    },
    password: {
      type: GraphQLString,
      resolve: person =>
        process.env.NODE_ENV !== "production" && person.password,
      description: "The users password, available for testing purpouses."
    },
    token: {
      type: GraphQLString,
      resolve: person => person.token,
      description: "A uniquely identifiable token."
    },
    personId: {
      type: GraphQLString,
      resolve: person => person.id,
      description: "A uniquely identifiable local id."
    },
    created: createdField(),
    edited: editedField(),
    id: globalIdField("Person")
  }),
  interfaces: () => [nodeInterface]
});

export default PersonType;
