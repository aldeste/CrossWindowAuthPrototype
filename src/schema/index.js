// @flow
import { GraphQLID, GraphQLObjectType, GraphQLSchema } from "graphql";
import { fromGlobalId } from "graphql-relay";

import { nodeField } from "./relayNode";
import personType from "./types/personType";
import { getObjectFromTypeAndId } from "./apiHelper";

type arguments = { personId?: ?string, id?: ?string };
type globalIdObejct = { type?: ?string, id?: ?string };

const Root = new GraphQLObjectType({
  name: "Root",
  fields: () => ({
    viewer: {
      type: personType,
      args: {
        id: { type: GraphQLID },
        personId: { type: GraphQLID }
      },
      resolve(_: *, args: arguments, viewer: Object): Object {
        const getter = (id: string) =>
          getObjectFromTypeAndId(personType, id, viewer);
        const { personId, id: globalId } = args;

        if (typeof personId === "string") {
          return getter(personId);
        }

        if (typeof globalId === "string") {
          const { id, type }: globalIdObejct = fromGlobalId(globalId);
          if (!id || id === "" || type === "") {
            throw new Error(`No valid ID extracted from ${globalId}`);
          }

          return getter(id);
        }
        throw new Error(`You must provide id or personId`);
      }
    },
    node: nodeField
  })
});

export default new GraphQLSchema({ query: Root });
