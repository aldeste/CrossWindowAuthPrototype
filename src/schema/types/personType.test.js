import schema from "../";
import personType from "./personType";
import { graphql } from "graphql";
import { createLoaders } from "../apiHelper";

beforeEach(async () => {
  console.log = jest.fn();
  await require("../../database").initializeDatabase();
});

describe("personType is a GraphQL type", () => {
  const rootValue = {};
  const context = { loaders: createLoaders() };

  it("Has the name Person", () => {
    expect(personType.name).toBe("Person");
  });

  it("should return on every field", async () => {
    const query = `query Test {
      person(personId: 5) {
        name
        birthYear
        eyeColor
        gender
        hairColor
        height
        mass
        skinColor
        token
        id
        homeworld {
          id
          name
        }
      }
    }`;

    const { data } = await graphql(schema, query, rootValue, context);
    expect(data).toMatchSnapshot();
  });

  it("should return on every field with id", async () => {
    const query = `query Test {
      person(id: "cGVvcGxlOjU=") {
        name
        birthYear
        eyeColor
        gender
        hairColor
        height
        mass
        skinColor
        token
        id
        homeworld {
          id
          name
        }
      }
    }`;

    const { data } = await graphql(schema, query, rootValue, context);
    expect(data).toMatchSnapshot();
  });

  it("should return nothing if no argument is supplied", async () => {
    const query = `query Test {
      viewer {
        name
      }
    }`;

    const { data } = await graphql(schema, query, rootValue, context);

    expect(data.viewer).toBe(null);
  });

  it("should return nothing if id argument is invalid", async () => {
    const query = `query Test {
      viewer(id: "e") {
        name
      }
    }`;

    const result = await graphql(schema, query, rootValue, context);

    expect(result.errors).toBeDefined();
  });

  it("should return nothing if personId argument is invalid", async () => {
    const query = `query Test {
      viewer(personId: e) {
        name
      }
    }`;

    const result = await graphql(schema, query, rootValue, context);

    expect(result.errors).toBeDefined();
  });
});
