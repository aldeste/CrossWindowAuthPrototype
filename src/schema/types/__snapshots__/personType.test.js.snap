// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`personType is a GraphQL type should return on every field 1`] = `
Object {
  "person": Object {
    "birthYear": "19BBY",
    "eyeColor": "brown",
    "gender": "female",
    "hairColor": "brown",
    "height": 150,
    "id": "cGVvcGxlOjU=",
    "mass": 49,
    "name": "Leia Organa",
    "skinColor": "light",
    "token": "cGVvcGxlOjU=",
  },
}
`;

exports[`personType is a GraphQL type should return on every field with id 1`] = `
Object {
  "person": Object {
    "birthYear": "19BBY",
    "eyeColor": "brown",
    "gender": "female",
    "hairColor": "brown",
    "height": 150,
    "id": "cGVvcGxlOjU=",
    "mass": 49,
    "name": "Leia Organa",
    "skinColor": "light",
    "token": "cGVvcGxlOjU=",
  },
}
`;
